### PR TITLE
Projectile refactor 1

### PR DESCRIPTION
**Pull request by MrRowey**
Batch one of Projectiles refactoring

### CODE REVIEW

File: lua/aeonprojectiles.lua [hunk 1]
```diff
@@ -136,6 +136,7 @@ ADepthChargeProjectile = Class(OnWaterEntryEmitterProjectile) {
     FxTrails = {'/effects/emitters/torpedo_munition_trail_01_emit.bp',},
     TrailDelay = 0,
     TrackTime = 0,
+    CountdownLength = 101,
```
Comment (Garanas) [inline lua/aeonprojectiles.lua:139]: What does this do?

File: projectiles/AANDepthCharge01/AANDepthCharge01_script.lua [hunk 2]
```diff
@@ -24,6 +24,13 @@ AANDepthCharge01 = Class(ADepthChargeProjectile) {
         end
     end,
 
+    OnEnterWater = function(self)
+        self:SetMaxSpeed(20)
```
Comment (Garanas) [inline projectiles/AANDepthCharge01/AANDepthCharge01_script.lua:28]: Feels to me this should also call the base function

File: projectiles/AANDepthCharge01/AANDepthCharge01_script.lua [hunk 3]
```diff
@@ -38,6 +45,7 @@ AANDepthCharge01 = Class(ADepthChargeProjectile) {
     end,
 
     OnImpact = function(self, TargetType, TargetEntity)
+        self.HasImpacted = true
```
Comment (Garanas) [inline projectiles/AANDepthCharge01/AANDepthCharge01_script.lua:48]: If this value is not used then there is no need to define it

File: projectiles/AANTorpedoCluster01/AANTorpedoCluster01_script.lua [hunk 4]
```diff
@@ -37,20 +21,16 @@ AANTorpedoCluster01 = Class(ATorpedoCluster) {
     OnEnterWater = function(self) 
         ATorpedoCluster.OnEnterWater(self)
 
-        -- create two child projectiles
+-- create two child projectiles
         for i = 0, 1 do
             proj = self:CreateChildProjectile('/projectiles/AANTorpedoClusterSplit01/AANTorpedoClusterSplit01_proj.bp' )
             proj:PassDamageData(self.DamageData)
```
Comment (Garanas) [inline projectiles/AANTorpedoCluster01/AANTorpedoCluster01_script.lua:27]: in this case you can also write: `proj.DamageData = self.DamageData`

File: lua/aeonprojectiles.lua (hunk 1 above)
Comment (Garanas) [reply lua/aeonprojectiles.lua:139]: I get it now, that is the count down timer. Is that function available at this scope?

File: projectiles/AIFFragmentationSensorShell01/AIFFragmentationSensorShell01_script.lua [hunk 5]
```diff
@@ -41,11 +41,11 @@ AIFFragmentationSensorShell01 = Class(AArtilleryFragmentationSensorShellProjecti
         -- Launch projectiles at semi-random angles away from split location
         for i = 0, numProjectiles - 1 do
             xVec = vx + (math.sin(angleInitial + (i*angle) + RandomFloat(-angleVariation, angleVariation))) * spreadMul
-            zVec = vz + (math.cos(angleInitial + (i*angle) + RandomFloat(-angleVariation, angleVariation))) * spreadMul 
+            zVec = vz + (math.cos(angleInitial + (i*angle) + RandomFloat(-angleVariation, angleVariation))) * spreadMul
             local proj = self:CreateChildProjectile(bp.FragmentId)
             proj:SetVelocity(xVec,yVec,zVec)
             proj:SetVelocity(velocity)
-            proj:PassDamageData(self.DamageData)                        
+            proj:PassDamageData(self.DamageData)
```
Comment (Garanas) [inline projectiles/AIFFragmentationSensorShell01/AIFFragmentationSensorShell01_script.lua:48]: You can also use `proj.DamageData = self.DamageData`, as you pass the damage data unchanged

File: projectiles/AIFGuidedMissile02/AIFGuidedMissile02_Script.lua [hunk 6]
```diff
@@ -8,11 +8,11 @@ local AGuidedMissileProjectile = import("/lua/aeonprojectiles.lua").AGuidedMissi
 AIFGuidedMissile02 = Class(AGuidedMissileProjectile) {
     OnCreate = function(self)
 		AGuidedMissileProjectile.OnCreate(self)
-		self:ForkThread( self.MovementThread )
+		self.Trash:Add(ForkThread( self.MovementThread))
     end,
 
 	MovementThread = function(self)
-		WaitSeconds(0.6)
+		WaitTicks(6)
```
Comment (Garanas) [inline projectiles/AIFGuidedMissile02/AIFGuidedMissile02_Script.lua:15]: should be `WaitTicks(7)`, if I am not mistaken.

File: projectiles/AIFMissileSerpentine01/AIFMissileSerpentine01_script.lua [hunk 7]
```diff
@@ -5,14 +5,14 @@ AIFMissileSerpentine01 = Class(AMissileSerpentineProjectile) {
...
-        self.MoveThread = self:ForkThread(self.MovementThread)
+        self.MoveThread = self.Trash:Add(ForkThread(self.MovementThread,self))
     end,
 
     MovementThread = function(self)        
         self.WaitTime = 0.1
         self.Distance = self:GetDistanceToTarget()
         self:SetTurnRate(8)
-        WaitSeconds(0.3)        
+        WaitTicks(3)
```
Comment (Garanas) [inline projectiles/AIFMissileSerpentine01/AIFMissileSerpentine01_script.lua:15]: Should be `WaitTicks(4)`

File: projectiles/AIFMissileSerpentine01/AIFMissileSerpentine01_script.lua [hunk 8]
```diff
@@ -23,20 +23,20 @@ AIFMissileSerpentine01 = Class(AMissileSerpentineProjectile) {
...
+            WaitTicks(21)
             self:SetTurnRate(10)
         elseif dist > 30 and dist <= 50 then
 						self:SetTurnRate(12)
-						WaitSeconds(1.5)
+						WaitTicks(12)
             self:SetTurnRate(12)
         elseif dist > 10 and dist <= 25 then
-            WaitSeconds(0.3)
+            WaitTicks(3)
```
Comment (Garanas) [inline projectiles/AIFMissileSerpentine01/AIFMissileSerpentine01_script.lua:39]: Should be `WaitTicks(4)`

File: projectiles/AIFMissileSerpentine01/AIFMissileSerpentine01_script.lua [hunk 9]
```diff
@@ -23,20 +23,20 @@ AIFMissileSerpentine01 = Class(AMissileSerpentineProjectile) {
...
         end
         if dist > 50 then        
             --Freeze the turn rate as to prevent steep angles at long distance targets
-            WaitSeconds(2)
+            WaitTicks(21)
             self:SetTurnRate(10)
         elseif dist > 30 and dist <= 50 then
 						self:SetTurnRate(12)
-						WaitSeconds(1.5)
+						WaitTicks(12)
```
Comment (Garanas) [inline projectiles/AIFMissileSerpentine01/AIFMissileSerpentine01_script.lua:36]: This should be `WaitTicks(16)`

**Review by Garanas — CHANGES_REQUESTED** [review]
Good start - let's tidy it up and get this merged!

File: lua/aeonprojectiles.lua (hunk 1 above)
Comment (MrRowey) [reply lua/aeonprojectiles.lua:139]: Didn't mean to add this here :) removed

File: projectiles/AANDepthCharge03/AANDepthCharge03_script.lua [hunk 10]
```diff
@@ -1,71 +1,46 @@
...
     CountdownExplosion = function(self)
-        WaitSeconds(self.CountdownLength)
-
+        WaitTicks(self.CountdownLength)
         if not self.HasImpacted then
             self:OnImpact('Underwater', nil)
         end
     end,
 
     OnEnterWater = function(self)
```
Comment (Garanas) [inline projectiles/AANDepthCharge03/AANDepthCharge03_script.lua:21]: Needs to call the base class I think

**Review by Garanas — CHANGES_REQUESTED** [review]
Almost there, great job 👍

File: projectiles/AANDepthCharge03/AANDepthCharge03_script.lua (hunk 10 above)
Comment (MrRowey) [reply projectiles/AANDepthCharge03/AANDepthCharge03_script.lua:21]: @Garanas this didn't call the base function to begin with so and their no 
```
ADepthChargeProjectile.CountdownExplosion(self)
```
to use

Comment (Garanas) [reply projectiles/AANDepthCharge03/AANDepthCharge03_script.lua:21]: We're talking here about `OnEnterWater` however

Comment (MrRowey) [reply projectiles/AANDepthCharge03/AANDepthCharge03_script.lua:21]: xD Oohhhhhhh yah i can fix that :)

**Review by Garanas — APPROVED** [review]

**Pull request merged**